File: server/leave-service/app/services/leave_service.py
from datetime import datetime
from typing import Optional, List
import uuid

from app.models.leave_request import LeaveRequestInDB, LeaveRequestOut


mock_leaves: List[LeaveRequestInDB] = []


def generate_id() -> str:
    return f"lr-{uuid.uuid4().hex[:8]}"


def get_leave_by_id(leave_id: str) -> Optional[LeaveRequestInDB]:
    return next((leave for leave in mock_leaves if leave.id == leave_id), None)


def get_leaves_by_user(user_id: str) -> List[LeaveRequestInDB]:
    return [leave for leave in mock_leaves if leave.userId == user_id]


def create_leave(user_id: str, start_time: datetime, end_time: datetime, reason: str) -> LeaveRequestOut:
    now = datetime.utcnow()
    leave = LeaveRequestInDB(
        id=generate_id(),
        userId=user_id,
        startTime=start_time,
        endTime=end_time,
        reason=reason,
        status='draft',
        workflowId=None,
        createdAt=now,
        updatedAt=now,
    )
    mock_leaves.append(leave)
    return LeaveRequestOut(**leave.model_dump())


def update_leave(leave_id: str, **updates) -> Optional[LeaveRequestOut]:
    leave = get_leave_by_id(leave_id)
    if not leave:
        return None

    for key, value in updates.items():
        if value is not None and hasattr(leave, key):
            setattr(leave, key, value)
    leave.updatedAt = datetime.utcnow()

    return LeaveRequestOut(**leave.model_dump())


def submit_leave(leave_id: str, workflow_id: str) -> Optional[LeaveRequestOut]:
    leave = get_leave_by_id(leave_id)
    if not leave:
        return None

    leave.status = 'submitted'
    leave.workflowId = workflow_id
    leave.updatedAt = datetime.utcnow()
    return LeaveRequestOut(**leave.model_dump())

File: client/src/services/leaveService.ts
import api from './api';
import { LeaveRequest } from '../types';

interface CreateLeaveRequest {
  startTime: string;
  endTime: string;
  reason: string;
}

export async function getLeaveList(): Promise<LeaveRequest[]> {
  const response = await api.get('/api/leave');
  return response.data.data;
}

export async function getLeaveById(id: string): Promise<LeaveRequest> {
  const response = await api.get(`/api/leave/${id}`);
  return response.data.data;
}

export async function createLeaveRequest(data: CreateLeaveRequest): Promise<LeaveRequest> {
  const response = await api.post('/api/leave', data);
  return response.data.data;
}

export async function updateLeave(id: string, data: Partial<LeaveRequest>): Promise<LeaveRequest> {
  const response = await api.put(`/api/leave/${id}`, data);
  return response.data.data;
}

export async function submitLeave(id: string): Promise<void> {
  await api.post(`/api/leave/${id}/submit`);
}